skiplist.go

```diff
@@ -173,3 +340,39 @@ func NewWithMaxLevel(maxLevel int) *SkipList {
 func New() *SkipList {
 	return NewWithMaxLevel(DefaultMaxLevel)
 }
+
+// CompactList compacts the skiplist by removing elements that are older than maxWait.
+// Returns number of sequences compacted and number of sequences left in list
+func (list *SkipList) CompactList(timeNow, maxWait int64) ([]SkippedSequenceEntry, int64, int64) {
+	list.mutex.Lock()
+	defer list.mutex.Unlock()
+
+	if list.Length == 0 {
+		// list is empty, nothing to compact
+		return nil, 0, 0
+	}
+
+	numCompacted := int64(0)
+	var compactedEntries []SkippedSequenceEntry
+	// iterate through bottom linked list to find elements that are older than maxWait
+	for c := list._front(); c != nil; c = c.Next() {
+		if (timeNow - c.key.Timestamp) >= maxWait {
+			prevs := list.getPrevElementNodes(c.Key())
+			numCompacted += c.key.GetNumSequencesInEntry()
+			// remove element
+			for k, v := range c.next {
+				prevs[k].next[k] = v
+			}
+			// alter node count and back elem if necessary
+			list.Length--
+			if c == list.backElem {
+				list.backElem = nil
+			}
+			// add to compacted entries
+			compactedEntries = append(compactedEntries, c.key)
+		}
+	}
+	// decrement the number of sequences in the list by the number of compacted sequences
+	list.NumSequencesInList -= numCompacted
+	return compactedEntries, numCompacted, list.NumSequencesInList
```

Comment:
`numCompacted` is now just `len(compactedEntries)` so I think you can replace the int and separate counting - less chance of bugs where these two diverge (e.g. early returns)